File: Algorithms/079.WordSearch.py
class Solution:
    # @param board, a list of lists of 1 length string
    # @param word, a string
    # @return a boolean
    def exist(self, board, word):
        def _dfs(board, word, m, n, level, i, j):
            if level == len(word): return True
            if i < 0 or j < 0 or i >= m or j >= n: return False
            if board[i][j] != word[level]: return False
            if visited[i][j]: return False
            visited[i][j] = True
            res = (_dfs(board, word, m, n, level+1, i-1, j)
                    or _dfs(board, word, m, n, level+1, i+1, j)
                    or _dfs(board, word, m, n, level+1, i, j-1)
                    or _dfs(board, word, m, n, level+1, i, j+1)
                    )
            visited[i][j] = False
            return res
        m, n = len(board), len(board[0])
        visited = [[False for i in range(n)] for j in range(m)]
        for i in range(m):
            for j in range(n):
                if _dfs(board, word, m, n, 0, i, j):
                    return True
        return False

File: Algorithms/cpp/079.WordSearch.cpp
class Solution {
public:
    // dfs
    bool exist(vector<vector<char>>& board, string word) {
        int m = board.size();
        int n = board[0].size();
        vector<vector<bool>> visited(m, vector<bool>(n, false));
        for (int i=0; i < m; ++i) {
            for (int j=0; j < n; ++j) {
                if (dfs(0, i, j, word, board, visited))
                    return true;
            }
        }
        return false;
    }
    bool dfs(int index, int i, int j, string& word, 
            vector<vector<char>>& board, vector<vector<bool>>& visited) {
        if (index == word.size())
            return true;
        if (i < 0 || j < 0 || i >= board.size() || j >= board[0].size())
            return false;
        if (board[i][j] != word[index])
            return false;
        if (visited[i][j])
            return false;
        visited[i][j] = true;
        bool result = dfs(index+1, i-1, j, word, board, visited) ||
            dfs(index+1, i+1, j, word, board, visited) ||
            dfs(index+1, i, j-1, word, board, visited) ||
            dfs(index+1, i, j+1, word, board, visited);
        visited[i][j] = false;
        return result;
    }
};
